plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTable.java

```diff
@@ -0,0 +1,137 @@
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v2Unixtime, 2), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndLongTimestamp()
+    {
+        assertQueryFails("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(1, 8),
+                format("Time travel data is not available for table %s. The requested time is beyond the allowed time travel period.", "tpch.\"test_iceberg_read_versioned_table\""));
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v1Unixtime, 8), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v2Unixtime, 8), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndShortTimestampWithTimezone()
+    {
+        assertQueryFails("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + shortTzTimestamp(1),
+                format("Time travel data is not available for table %s. The requested time is beyond the allowed time travel period.", "tpch.\"test_iceberg_read_versioned_table\""));
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + shortTzTimestamp(v1Unixtime), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + shortTzTimestamp(v2Unixtime), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndLongTimestampWithTimezone()
+    {
+        assertQueryFails("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + longTzTimestamp(1),
+                format("Time travel data is not available for table %s. The requested time is beyond the allowed time travel period.", "tpch.\"test_iceberg_read_versioned_table\""));
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + longTzTimestamp(v1Unixtime), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + longTzTimestamp(v2Unixtime), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testEndVersionInTableNameAndForClauseShouldFail()
+    {
+        assertQueryFails("SELECT * FROM \"test_iceberg_read_versioned_table@" + v1SnapshotId + "\" FOR VERSION AS OF " + v1SnapshotId,
+                "Cannot specify end version both in table name and FOR clause");
+
+        assertQueryFails("SELECT * FROM \"test_iceberg_read_versioned_table@" + v1SnapshotId + "\" FOR TIMESTAMP AS OF " + longTzTimestamp(v1Unixtime),
+                "Cannot specify end version both in table name and FOR clause");
+    }
+
+    private String shortTzTimestamp(long epochSeconds)
+    {
+        return "from_unixtime(" + epochSeconds + ")";
```

Comment:
`from_unixtime` semantics are never obvious to me (eg is it session zone dependent?)

Make the code more explicit

```java
private String timestampLiteral(long epochSeconds, int precision)
{
    return DateTimeFormatter.ofPattern("'TIMESTAMP '''uuuu-MM-dd HH:mm:ss." + "S".repeat(precision) + " VV''")
            .format(Instant.ofEpochMilli(epochSeconds).atZone(UTC));
}
```

this also frees you from needing to wrap the values in CAST and differentiate between "short" and "long".